### DISCUSSION TITLE
getServerSideProps locale & defaultLocale are undefined

### DISCUSSION THREAD

**Discussion by sashberd** [vercel/next.js#24718] [Help] [answered]
Hi
As mention in [documentation](https://nextjs.org/docs/basic-features/data-fetching#getserversideprops-server-side-rendering), getServerSideProps should expose locale and defaultLocale props via context. In my app I did as in documentations example, but  locale and defaultLocale are undefined:

next-18next.config.js

```
module.exports = {
  i18n: {
    defaultLocale: 'pl',
    locales: ['pl', 'en'],
  },
  preload: ['en', 'pl'],
  defaultNS: 'common',
  fallbackNS: 'common',
  fallbackLng: 'pl',
  debug: false,
  react: {
    useSuspense: false,
    wait: false,
  },
};
```
next.config.js: 

```
const { i18n } = require('./next-i18next.config')
module.exports = { i18n }
```

pages/project.js
```
export async function getServerSideProps(context) {
  const isRedirect = await redirectIfOldIdOrProjectNotVisible(context);
  if (isRedirect) {
    return {
      redirect: {
        destination: '/',
        statusCode: 302,
      },
    };
  }
  return {
    props: {
      ...(await serverSideTranslations(context.locale || context.defaultLocale, [
        'common',
        'header',
        'project',
        'footer',
      ])),
    },
  };
```

<img width="1040" alt="Screen Shot 2021-05-02 at 18 20 48" src="https://user-images.githubusercontent.com/11975561/116818275-342ec080-ab73-11eb-9a7b-b9dab7701084.png">

 "next": "^10.0.9",
 "next-i18next": "^8.2.0",
 "i18next": "^20.1.0",
 
 I broke my head, but still no luck. Anyone has asme issue and could help to solve it?
 Thanks

**Comment by bobarros** [comment]
I can't see any problem with the code you're presenting here. I was so curious that I copy and pasted them to test myself.

Didn't get any errors. Are you sure that you're getting undefined from ctx?

What do you get in your console when you try to print them?

```
console.log(context, context.locale, context.locales, context.defaultLocale);
```

**Reply by sashberd** [reply]
@bobarros Thanks for the help
The console is: 
context - is the huge object
locale - undefined,
defaultLocale - undefined
locales - [en, pl]
<img width="979" alt="Screen Shot 2021-05-03 at 14 17 39" src="https://user-images.githubusercontent.com/11975561/116869950-5596b780-ac1a-11eb-9245-67afca1e9cba.png">

**Reply by sashberd** [reply]
@bobarros Another question: Could those values be overwritten by some other values in _document or _app or any other previous rendered components

**Reply by bobarros** [reply]
Locale's server-side and _document or _app should not mess with them for no reason. That's why some people want to change default local and can't --> https://github.com/vercel/next.js/discussions/18419

Still a mystery to me what's going on.

One thing that I notice it's that you have a typo when you wrote your file name here -> next-18next.config.js
You're missing the "i" -> next-i18next.config.js
By any chance, are you doing this somewhere else without throwing an error? Did you copy and paste this name? We never know, right?

One other thing that I would try, it's to force default language and debug to see what happens:

```javascript
return {
    props: {
      ...(await serverSideTranslations(context.locale || "pl", [
        'common',
        'header',
        'project',
        'footer',
      ])),
    }
```

**Reply by sashberd** [reply]
As for the file name, it is a type in the question. My file names are good:
<img width="402" alt="Screen Shot 2021-05-03 at 15 58 22" src="https://user-images.githubusercontent.com/11975561/116879695-a06bfb80-ac29-11eb-9adf-375b999eb9bc.png">

As for the forcing default language:
I added next logs:
console.log('!!!!!I AM IN getServerSideProps', context.locale);
console.log('!!!!!I AM IN getServerSideProps', context.locale || 'pl');
The log was always:
!!!!!I AM IN getServerSideProps undefined
!!!!!I AM IN getServerSideProps pl

**Reply by bobarros** [reply]
But when you force it, everything else works as it should? No errors?

**Reply by sashberd** [reply]
Yes, all works. Page is loading and no errors

**Comment by sashberd** [comment]
@bobarros Any chance that custom server and the middlwares in it could overwrite request values?

**Reply by bobarros** [reply]
After some research and our conversation, I realize that a custom server with express doesn't send the information you need as default. It's not a bug of your application.

You have these two options below and some more. In the end, you can change in many ways.
I'm reading the req language and sending response as needed.

```jsx

//// This is the one I would use it. If main user language it's 
//// not one of your options, use the one you want. In this case, hebrew.
export async function getServerSideProps(context) {
  const isRedirect = await redirectIfOldIdOrProjectNotVisible(context);

  if (isRedirect) {
    return {
      redirect: {
        destination: '/',
        statusCode: 302,
      },
    };
  }
  
  const preferLanguage = context.req.headers["accept-language"].substr(0,2);
  const outputLanguage = context.locales.includes(preferLanguage) ? preferLanguage : "he";

  return {
    props: {
      ...(await serverSideTranslations(outputLanguage || "he", [
        'common',
        'header',
        'project',
        'footer',
      ])),
    },
  };
}


//// This is another option if you want to check all browsers languages available
export async function getServerSideProps(context) {
  const isRedirect = await redirectIfOldIdOrProjectNotVisible(context);

  if (isRedirect) {
    return {
      redirect: {
        destination: '/',
        statusCode: 302,
      },
    };
  }

  const allLanguage = context.req.headers["accept-language"];
  const outputLanguage = context.locales.find(locale => {
    return allLanguage.includes(locale);
  });

  return {
    props: {
      ...(await serverSideTranslations(outputLanguage || "he", [
        'common',
        'header',
        'project',
        'footer',
      ])),
    },
  };
}
```

**Reply by sashberd** [reply]
Thanks Bruno, I will use your solution
So, in other words, nextJS with custom service does not support getServerSideProps with locales params. Maybe ask NextJS developers team to provide such approach?

**Reply by bobarros** [reply]
I tried this custom server that they use as an example and got no locales params also. 
Maybe it's something to discuss with them. I will investigate some more, before opening the issue.

--> https://nextjs.org/docs/advanced-features/custom-server

All the best.

**Reply by bobarros** [reply]
They already have an issue about it: https://github.com/vercel/next.js/issues/20761

**Reply by sashberd** [reply]
@bobarros Thanks for your help!!!!

**Comment by sashberd** [comment, marked as answer]
For those who are looking for the workaround:

First of all add next keys to each query params object:
```
__nextLocale: req.params.locale,
__nextDefaultLocale: 'he',
```

e.g:
```
app.render(req, res, pagePath, {
          ...queryParams,
          __nextLocale: req.params.locale,
          __nextDefaultLocale: 'he',
        });
```

Then, for each server.get url  append next optional param: `  server.get('/:locale?/project/:id', (req, res) => {...}`

Thus, the approach above passed locale data from a custom server and you can use it with routes.

**Reply by aNyMoRe0505** [reply]
thanks!!

**Reply by Rechdan** [reply]
Working for now, thanks!!

**Comment by krzysiek1507** [comment]
Hey @samcx. It looks like this stopped working. Are you aware if that was intentional?

**Reply by samcx** [reply]
@krzysiek1507 Not that I'm aware of. Are you able to pin this to a canary?

**Reply by krzysiek1507** [reply]
Somewhere in 13.x `__nextLocale`, `__nextDefaultLocale` and locale started being pushed down and back to the client URL. At the same time the `req.url` changed from 12.x `/fr` to `/?locale=&__nextLocale=en-gb&__nextDefaultLocale=en-gb`. Our URLs are like: `example.com/:locale`

We don't use App router and I wonder if there were some changes in 13.x that would break i18n?

**Reply by krzysiek1507** [reply]
Did maybe something in `app.render` change?

**Reply by samcx** [reply]
@krzysiek1507 Not that I'm aware of (we didn't make any breaking changes with regards to internationalization in Pages Router). If recommend creating a new [bug report](https://github.com/vercel/next.js/issues/new?assignees=&labels=template:+bug&projects=&template=1.bug_report.yml) to report this!
